File: node/src/memory.ts
/**
 * Memory namespace for the Forge client.
 */

import type { Forge } from "./client.js";

export interface MemorySaveOptions {
  tags?: string[];
  metadata?: Record<string, unknown>;
}

export class MemoryNamespace {
  constructor(private client: Forge) {}

  async save(text: string, opts: MemorySaveOptions = {}): Promise<unknown> {
    const args: Record<string, unknown> = { text };
    if (opts.tags) args.tags = opts.tags;
    if (opts.metadata) args.metadata = opts.metadata;
    return this.client.toolsCall("memory_save", args);
  }

  async search(query: string, limit = 10): Promise<unknown> {
    return this.client.toolsCall("memory_search", { query, limit });
  }

  async list(limit = 50, offset = 0): Promise<unknown> {
    return this.client.toolsCall("memory_list", { limit, offset });
  }

  async delete(id: string): Promise<unknown> {
    return this.client.toolsCall("memory_delete", { id });
  }
}

File: rust/src/memory.rs
//! Memory namespace helpers.

use serde_json::{json, Value};

use crate::client::Forge;
use crate::error::Result;

pub struct Memory<'a> {
    client: &'a Forge,
}

impl<'a> Memory<'a> {
    pub(crate) fn new(client: &'a Forge) -> Self {
        Self { client }
    }

    pub fn save(&self, text: &str, tags: Option<&[&str]>) -> Result<Value> {
        let mut args = json!({ "text": text });
        if let Some(t) = tags {
            args["tags"] = json!(t);
        }
        self.client.tools_call("memory_save", args)
    }

    pub fn search(&self, query: &str, limit: u32) -> Result<Value> {
        self.client
            .tools_call("memory_search", json!({ "query": query, "limit": limit }))
    }

    pub fn list(&self, limit: u32, offset: u32) -> Result<Value> {
        self.client
            .tools_call("memory_list", json!({ "limit": limit, "offset": offset }))
    }

    pub fn delete(&self, id: &str) -> Result<Value> {
        self.client.tools_call("memory_delete", json!({ "id": id }))
    }
}

impl Forge {
    pub fn memory(&self) -> Memory<'_> {
        Memory::new(self)
    }
}
